scripts/run_inference.py

```diff
@@ -0,0 +1,234 @@
+                The subdictionary of adapter_config, containing the keys in overrides,
+                with the values that we are going to replace.
+        """
+        # For now, we only expect to patch the base model; this may change in the future,
+        # but ensure that anything we are patching is defined in the original config
+        if not set(overrides.keys()).issubset(set(adapter_config.keys())):
+            raise KeyError("Adapter config overrides must be set in the config being patched")
+        return {key: adapter_config[key] for key in overrides}
+
+    def __enter__(self):
+        """Apply the config overrides and saved the patched values."""
+        self.patched_values = self._apply_config_changes(self.overrides)
+
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        """Apply the patched values over our exported overrides."""
+        self._apply_config_changes(self.patched_values)
+
+
+### Funcs for loading and running models
+class TunedCausalLM:
+    def __init__(self, model, tokenizer, device):
+        self.peft_model = model
+        self.tokenizer = tokenizer
+        self.device = device
+
+    @classmethod
+    def load(cls, checkpoint_path: str, base_model_name_or_path: str=None) -> "TunedCausalLM":
+        """Loads an instance of this model.
+
+        Args:
+            checkpoint_path: str
+                Checkpoint model to be loaded, which is a directory containing an
+                adapter_config.json.
+            base_model_name_or_path: str [Default: None]
+                Override for the base model to be used.
+
+        By default, the paths for the base model and tokenizer are contained within the adapter
+        config of the tuned model. Note that in this context, a path may refer to a model to be
+        downloaded from HF hub, or a local path on disk, the latter of which we must be careful
+        with when using a model that was written on a different device.
+
+        Returns:
+            TunedCausalLM
+                An instance of this class on which we can run inference.
+        """
+        overrides = {"base_model_name_or_path": base_model_name_or_path} if base_model_name_or_path is not None else {}
+        tokenizer = AutoTokenizer.from_pretrained(checkpoint_path)
+        # Apply the configs to the adapter config of this model; if no overrides
+        # are provided, then the context manager doesn't have any effect.
+        with AdapterConfigPatcher(checkpoint_path, overrides):
+            try:
+                peft_model = AutoPeftModelForCausalLM.from_pretrained(checkpoint_path)
+            except OSError as e:
+                print("Failed to initialize checkpoint model!")
+                raise e
+        device = "cuda" if torch.cuda.is_available() else None
+        print(f"Inferred device: {device}")
+        peft_model.to(device)
+        return cls(peft_model, tokenizer, device)
+
+
+    def run(self, text: str, *, max_new_tokens: int) -> str:
```

Comment:
we arent doing anything with the extra arguments * . maybe it is a typo here?

Reply:
It's kind of an obscure feature, but `*` in the func def forces everything after it to be passed as keyword arguments, and they can't be passed positionally.

```
>>> def foo(bar, *, baz):
...     pass
...
>>>
>>> foo(3, 100) # can't pass baz as a positional arg
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: foo() takes 1 positional argument but 2 were given
>>> foo(bar=3, baz=100) # as long as we pass kwarg here, it's fine
```

Since there are lots of options that can be used when generating, I thought this would be best to avoid confusion in the future

Reply:
oh so it was on purpose :) thank you Alex !